File: Minimum BST Sum Subtree.java
class Solution {
    static int ans = Integer.MAX_VALUE; //track of the minimum size of the subtree
    
    public static int minSubtreeSumBST(int target, Node root) {
        ans = Integer.MAX_VALUE; // reset the inctance variable
        find(target, root); // find the minimum size of the subtree
        return ans == Integer.MAX_VALUE ? -1 : ans; // if ans is still Integer.MAX_VALUE, it means no subtree was found
    }
    
    public static void find(int target, Node root) {
        if (root == null) return; // base case
        
        int sum = sum(root); // calculate the sum of the current subtree
        if (sum == target && isBST(root)) { // if the sum is equal to target and the current subtree is a BST
            ans = Math.min(ans, size(root)); // update the global variable with the minimum size
        }
        
        find(target, root.left); // recursively find in left subtree
        find(target, root.right); // recursively find in right subtree
    }
    
    private static int sum(Node root) { // helper method to calculate the sum of a subtree
        if (root == null) return 0;
        return root.data + sum(root.left) + sum(root.right);
    }
    
    private static int size(Node root) { // helper method to calculate the size of a subtree
        if (root == null) return 0;
        return 1 + size(root.left) + size(root.right);
    }
    
    private static boolean isBST(Node root) { // helper method to check if a subtree is a BST
        return isBSTUtil(root, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }
    
    private static boolean isBSTUtil(Node node, int min, int max) { // utility method for checking if a subtree is a BST
        if (node == null) return true;
        
        if (node.data < min || node.data > max) return false;
        
        return isBSTUtil(node.left, min, node.data - 1)
            && isBSTUtil(node.right, node.data + 1, max);
    }
}

File: Minimum BST Sum Subtree.cpp
class Solution {
public:
    int ans=INT_MAX;
    int tar;
    
    //  <node,<sum,<flag,<min,max>>>>
    pair<int,pair<int,pair<bool,pair<int,int>>>> dfs(Node* node){
        bool flag=true;
        int mn=node->data, mx=node->data;
        int sum=node->data;
        int nodes=1;
        
        if(node->left){
            pair<int,pair<int,pair<bool,pair<int,int>>>> left=dfs(node->left);
            sum+=left.second.first;
            nodes+=left.first;
            if( left.second.second.first && (left.second.second.second.second < node->data) ){
                mn=left.second.second.second.first;
            } else {
                flag=false;
            }
        }
        if(node->right){
            pair<int,pair<int,pair<bool,pair<int,int>>>> right=dfs(node->right);
            sum+=right.second.first;
            nodes+=right.first;
            if( right.second.second.first && (right.second.second.second.first > node->data) ){
                mx=right.second.second.second.second;
            } else {
                flag=false;
            }
        }

        if(flag && sum==tar){
            ans=min(ans, nodes);
        }
        return {nodes,{sum,{flag,{mn,mx}}}};
    }
    
    int minSubtreeSumBST(int target, Node *root) {
        tar=target;
        dfs(root);
        return ans==INT_MAX?-1:ans;
    }
};
